host_reference/domain_wall_dslash_reference.cpp

```diff
@@ -975,6 +1235,216 @@ void mdw_matpc(void *out, void **gauge, void *in, double _Complex *kappa_b, doub
+            (char *)out + precision * Vh * spinor_site_size * xs, Vh * spinor_site_size, precision);
+    }
+  } else if (symmetric && dagger) {
+    mdw_eofa_m5inv(tmp, in, parity[1], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    dslash_4_4d(out, gauge, tmp, parity[0], dagger, precision, gauge_param, mferm);
+    mdw_dslash_4_pre(tmp, gauge, out, parity[0], dagger, precision, gauge_param, mferm, b5, c5, true);
+    mdw_eofa_m5inv(out, tmp, parity[0], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    dslash_4_4d(tmp, gauge, out, parity[1], dagger, precision, gauge_param, mferm);
+    mdw_dslash_4_pre(out, gauge, tmp, parity[1], dagger, precision, gauge_param, mferm, b5, c5, true);
+    for (int xs = 0; xs < Ls; xs++) {
+      cxpay((char *)in + precision * Vh * spinor_site_size * xs, kappa2[xs],
+            (char *)out + precision * Vh * spinor_site_size * xs, Vh * spinor_site_size, precision);
+    }
+  } else if (!symmetric && !dagger) {
+    mdw_dslash_4_pre(out, gauge, in, parity[1], dagger, precision, gauge_param, mferm, b5, c5, true);
+    dslash_4_4d(tmp, gauge, out, parity[0], dagger, precision, gauge_param, mferm);
+    mdw_eofa_m5inv(out, tmp, parity[1], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    mdw_dslash_4_pre(tmp, gauge, out, parity[0], dagger, precision, gauge_param, mferm, b5, c5, true);
+    dslash_4_4d(out, gauge, tmp, parity[1], dagger, precision, gauge_param, mferm);
+    mdw_eofa_m5(tmp, in, parity[0], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    for (int xs = 0; xs < Ls; xs++) {
+      cxpay((char *)tmp + precision * Vh * spinor_site_size * xs, kappa2[xs],
+            (char *)out + precision * Vh * spinor_site_size * xs, Vh * spinor_site_size, precision);
+    }
+  } else if (!symmetric && dagger) {
+    dslash_4_4d(out, gauge, in, parity[0], dagger, precision, gauge_param, mferm);
+    mdw_dslash_4_pre(tmp, gauge, out, parity[1], dagger, precision, gauge_param, mferm, b5, c5, true);
+    mdw_eofa_m5inv(out, tmp, parity[0], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    dslash_4_4d(tmp, gauge, out, parity[1], dagger, precision, gauge_param, mferm);
+    mdw_dslash_4_pre(out, gauge, tmp, parity[0], dagger, precision, gauge_param, mferm, b5, c5, true);
+    mdw_eofa_m5(tmp, in, parity[0], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    for (int xs = 0; xs < Ls; xs++) {
+      cxpay((char *)tmp + precision * Vh * spinor_site_size * xs, kappa2[xs],
+            (char *)out + precision * Vh * spinor_site_size * xs, Vh * spinor_site_size, precision);
+    }
+  } else {
+    errorQuda("Unsupported matpc_type=%d dagger=%d", matpc_type, dagger);
+  }
+
+  free(tmp);
+}
+
+// helper for creating extended gauge fields
+static cpuGaugeField *create_extended_gauge(void **gauge, QudaGaugeParam &gauge_param, const int *R)
```

Comment:
Won't hold the PR up on this, but we should put this routine in a file in `utils/`. @hummingtree if you could file a bug (well, feature request) for this I'd be grateful.

Reply:
I will add this to `utils/`.